ngraph/core/reference/include/ngraph/runtime/reference/binary_convolution.hpp

```diff
@@ -0,0 +1,203 @@
+                            for (int i_x = -p.pads_begin[2];
+                                 i_x <= (p.pads_end[2] + input_size_x - dilated_filter_size_x);
+                                 i_x += p.strides[2])
+                            {
+                                auto input_channel = batch;
+                                auto filter_channel = filter;
+                                T_IN sum = 0;
+                                size_t filter_channels_count = filter_shape[0];
+                                while (filter_channels_count--)
+                                {
+                                    for (int f_z = 0; f_z < filter_size_z; ++f_z)
+                                    {
+                                        for (int f_y = 0; f_y < filter_size_y; ++f_y)
+                                        {
+                                            for (int f_x = 0; f_x < filter_size_x; ++f_x)
+                                            {
+                                                int rel_i_z = i_z + (f_z * p.dilation[0]);
+                                                int rel_i_y = i_y + (f_y * p.dilation[1]);
+                                                int rel_i_x = i_x + (f_x * p.dilation[2]);
+
+                                                bool padding =
+                                                    !(in_range(rel_i_x, {0, input_size_x}) &&
+                                                      in_range(rel_i_y, {0, input_size_y}) &&
+                                                      in_range(rel_i_z, {0, input_size_z}));
+
+                                                int f_buf_idx =
+                                                    (f_z * filter_size_y * filter_size_x) +
+                                                    (f_y * filter_size_x) + f_x;
+                                                int i_buf_idx =
+                                                    (rel_i_z * input_size_y * input_size_x) +
+                                                    (rel_i_y * input_size_x) + rel_i_x;
+
+                                                T_IN in_val = padding
+                                                                  ? static_cast<T_IN>(pad_value)
+                                                                  : static_cast<T_IN>(
+                                                                        input_channel[i_buf_idx]);
+
+                                                T_F f_val = extract_bit(
+                                                    filter_channel[f_buf_idx / n_bits],
+                                                    (n_bits - 1) - (f_buf_idx % n_bits));
+
+                                                if (in_val == static_cast<T_IN>(f_val))
+                                                {
+                                                    sum += static_cast<T_IN>(1);
+                                                }
```

Comment:
Please refactor this computation and use names from spec (e.g. xnor, popcount, bit_count). It'll be easier to understand what is going on.

`
const bit_count = static_cast<T_IN>(filter_channel_size);
inline T xnor(T a, T b) { return a == b; } 
 
if (xnor(in_val, f_val)) {
  popcount ++;
}

*out = (2 * popcount- bit_count);
`

Reply:
Done